interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,469 @@
+// have been deemed safe and have been vetted by the security team.
+var allowedMountOptions = []string{
+	"async",
+	"atime",
+	"bind",
+	"diratime",
+	"dirsync",
+	"iversion",
+	"lazytime",
+	"nofail",
+	"noiversion",
+	"nomand",
+	"noatime",
+	"nodev",
+	"nodiratime",
+	"noexec",
+	"nolazytime",
+	"norelatime",
+	"nosuid",
+	"nostrictatime",
+	"nouser",
+	"relatime",
+	"strictatime",
+	"sync",
+	"ro",
+	"rw",
+}
+
+// A few mount flags are special in that if they are specified, the filesystem
+// type is ignored. We list them here, and we will ensure that the plug
+// declaration does not specify a type, if any of them is present among the
+// options.
+var optionsWithoutFsType = []string{
+	"bind",
+	// Note: the following flags should also fall into this list, but we are
+	// not currently allowing them (and don't plan to):
+	// - "make-private"
+	// - "make-shared"
+	// - "make-slave"
+	// - "make-unbindable"
+	// - "move"
+	// - "remount"
+}
+
+// List of allowed filesystem types. This can be extended, keeping in mind that
+// the filesystems in the following list were considered either dangerous or
+// not relevant for this interface:
+//   bpf
+//   cgroup
+//   cgroup2
+//   debugfs
+//   devpts
+//   ecryptfs
+//   hugetlbfs
+//   overlayfs
+//   proc
+//   securityfs
+//   sysfs
+//   tracefs
+var allowedFSTypes = []string{
+	"aufs",
+	"autofs",
+	"btrfs",
+	"ext2",
+	"ext3",
+	"ext4",
+	"hfs",
+	"iso9660",
+	"jfs",
+	"msdos",
+	"ntfs",
+	"ramfs",
+	"reiserfs",
+	"squashfs",
+	"tmpfs",
+	"ubifs",
+	"udf",
+	"ufs",
+	"vfat",
+	"zfs",
+	"xfs",
+}
+
+// mountControlInterface allows creating transient and persistent mounts
+type mountControlInterface struct {
+	commonInterface
+}
+
+// The "what" and "where" attributes end up in the AppArmor profile, surrounded
+// by double quotes; to ensure that a malicious snap cannot inject arbitrary
+// rules by specifying something like
+//   where: $SNAP_DATA/foo", /** rw, #
+// which would generate a profile line like:
+//   mount options=() "$SNAP_DATA/foo", /** rw, #"
+// (which would grant read-write access to the whole filesystem), it's enough
+// to exclude the `"` character: without it, whatever is written in the
+// attribute will not be able to escape being treated like a pattern.
+//
+// To be safe, there's more to be done: the pattern also needs to be valid, as
+// a malformed one (for example, a pattern having an unmatched `}`) would cause
+// apparmor_parser to fail loading the profile. For this situation, we use the
+// PathPattern interface to validate the pattern.
+//
+// Besides that, we are also excluding the `@` character, which is used to mark
+// AppArmor variables (tunables): when generating the profile we lack the
+// knowledge of which variables have been defined, so it's safer to exclude
+// them.
+var (
+	whatRegexp  = regexp.MustCompile(`^/[^"@]*$`)
+	whereRegexp = regexp.MustCompile(`^(/media|\$SNAP_COMMON|\$SNAP_DATA)/[^"@]*$`)
```

Comment:
why /media ?  I think in principle we want to allow / and control what is perimtted in terms of absolute targets in the snap-declaration?

Reply:
shouldn't we also prevent $ from appearing multiple times?

Reply:
@pedronis I assume by "allow / and ..." you mean to allow arbitrary absolute paths and not to allow mounting something at "/" exactly which would be very weird and break many things?

Assuming you meant absolute paths, maybe where should be something like:

```go
	whereRegexp = regexp.MustCompile(`^(/[^\$"@]+|\$SNAP_DATA|\$SNAP_COMMON)[^\$"@]*$`)
```

which will disallow "/" or any other string starting with an env var (except `$SNAP_DATA` or `$SNAP_COMMON`), but otherwise allow anything underneath `$SNAP_DATA` or `$SNAP_COMMON`, as well as disallowing any other env vars in the string, any `@` or any `"`. It also disallows `$SNAP_DATA/$SNAP_DATA`

Reply:
yes, I meant any absolute path (at this level)

Reply:
The dollar sign is not dangerous (AppArmor wouldn't treat it specially), but it sounds so suspicious that I think it's fine to disallow it.

I would suggest using a slightly simpler regexp:

	whereRegexp = regexp.MustCompile(`^(\$SNAP_COMMON|\$SNAP_DATA)?/[^\$"@]+$`)

The only practical difference from Ian's is that this disallows having just `$SNAP_COMMON` or `$SNAP_DATA` as the target, but requires `$SNAP_DATA/something`, which looks a reasonable restriction to me.

Reply:
I can see a use case behind `where: $SNAP_DATA` where a snap wants it's entire data dir to be located on some other partition than the main writable partition on it, but I suppose we could enable that use case when folks ask for it. 

@pedronis what do you think of @mardy's proposal?

Reply:
I'm fine with the  simplified proposal.